src/channel.cr

```diff
@@ -653,4 +653,65 @@ class Channel(T)
       raise ClosedError.new
     end
   end
+
+  class TimeoutAction
+    include SelectAction(Nil)
+
+    # Total amount of time to wait
+    @timeout : Time::Span
+
+    def initialize(@timeout : Time::Span)
+    end
+
+    def execute : DeliveryState
+      Fiber.current.timed_out = false
+      DeliveryState::None
+    end
+
+    def result : Nil
+      nil
+    end
+
+    def wait(context : SelectContext(Nil))
+      Fiber.timeout @timeout
+      context.try_trigger
```

Comment:
If I understand correctly, the below code will not receive the data during waiting.
```crystal
select
when ch.receive
when timeout(1.seconds)
end
```
Because the `context` is triggered before `Crystal::Scheduler.reschedule`, any `ch.send` will not awake the select fiber.

The timeline explanation:
```
    context.try_trigger( by timeout_action )
              |
              V
---->[ each action.wait ]---->[ reschedule ]------>[ ch.send ]---->[ timeout ]-->[ select fiber awake ]
                                    ^                   ^               ^
                                    |                   |               |
                              start waiting <-- 0.5s -->x<---- 0.5s --->x
                                                        |
                                            context.try_trigger ( fail because it has been triggered. )
```

I think the timeout worked as below is better.

```
                                                        context.try_trigger( failed then not enqueue fiber )
                                                                        |
                                                                        V
---->[ each action.wait ]---->[ reschedule ]------>[ ch.send ]---->[ timeout ]-->[ select fiber awake ]
                                    ^                   ^               ^
                                    |                   |               |
                              start waiting <-- 0.5s -->x<---- 0.5s --->x
                                                        |
                                            context.try_trigger ( succeed then enqueue fiber )

Or
                                                                      cleanup timeout
                                                                            |
                                                                            V
---->[ each action.wait ]---->[ reschedule ]------>[ ch.send ]---->[ select fiber awake ]---->x--->
                                    ^                   ^               ^                     ^
                                    |                   |               |                     |
                              start waiting <-- 0.5s -->x<--------------+--- 0.5s ----------->x
                                                        |
                                            context.try_trigger ( succeed then enqueue fiber )
```

Reply:
@firejox good catch. I'm surprised neither of the specs complains. I think somehow I (wrongly) expected to `Fiber.timeout` pause the execution of the fiber and continue from there.

I added more specs and move the `try_trigger` in the last commit.

The only specs that fails without the `try_trigger` change is "allows receiving while waiting (same fiber multiple times)"

Reply:
@firejox You are on fire! 👏

Reply:
@bcardiff I would suggest to put `context.try_trigger` into the event of `event loop`, just like [this](https://github.com/firejox/crystal/commit/31b1835a089aee6b12be134c1f3e34cf0770fcd3). Because it is hard to proof that `event loop` will not run after `ch.send`. If  `event loop` runs, the fiber will be enqueued again. Hence, I think that  `context.try_trigger` in event will be better.

Reply:
@firejox I fail to see a real difference with the proposed refactor. I think there is still no guarantee that if the timeout event is triggered and at the same time a channel-action can be performed which one will succeed the `try_trigger`. And that is fine, is the same scenario as competing channel-actions.

And on the downside, the fiber and the channels end up more coupled after the refactor IMO.

Reply:
@bcardiff Sorry for confusing refactor. The `try_trigger` determines which action can enqueue the fiber. However, the timeout event [here](https://github.com/crystal-lang/crystal/pull/8506/files#diff-4501d78495e747fe23a7fee036475ae2R35-R37) enqueue the fiber without `try_trigger` limitation. I worry that would lead the problem like #8300 . It enqueues the fiber more than one time.